addon/services/intl.js

```diff
@@ -37,205 +38,226 @@ function formatterProxy(formatType) {
-      const missingMessage = this.owner.resolveRegistration('util:intl/missing-message');
+      let missingMessage = this._owner.resolveRegistration('util:intl/missing-message');
 
       return missingMessage.call(this, key, localeNames);
     }
 
     return translation;
   },
 
+  /* @public */
   t(key, ...args) {
-    const [ options ] = args;
-    const translation = this.lookup(key, options && options.locale);
+    let [ options ] = args;
+    let translation = this.lookup(key, options && options.locale);
 
     return this.formatMessage(translation, ...args);
   },
 
+  /* @public */
   exists(key, localeName) {
-    const localeNames = this._localeWithDefault(localeName);
-    const adapter = get(this, 'adapter');
-
+    let localeNames = this.localeWithDefault(localeName);
     assert(`[ember-intl] locale is unset, cannot lookup '${key}'`, Array.isArray(localeNames) && localeNames.length);
 
-    return localeNames.some((localeName) => {
-      return adapter.has(localeName, key);
-    });
+    let adapter = get(this, 'adapter');
+
+    return localeNames.some(localeName => adapter.has(localeName, key));
   },
 
-  /**
-  * A utility method for registering CLDR data for
-  * intl-messageformat and intl-relativeformat.  This data is derived
-  * from formatjs-extract-cldr-data
-  *
-  * @method addLocaleData
-  * @param {Object} locale data
-  * @public
-  */
+  /* @public */
   addLocaleData(data) {
     IntlMessageFormat.__addLocaleData(data);
     IntlRelativeFormat.__addLocaleData(data);
   },
 
+  /* @public */
   addTranslation(localeName, key, value) {
     return this.localeFactory(localeName).addTranslation(key, value);
   },
 
+  /* @public */
   addTranslations(localeName, hash) {
     return this.localeFactory(localeName).addTranslations(hash);
   },
 
+  /* @public */
   setLocale(localeName) {
     if (!localeName) { return; }
 
-    const proposed = makeArray(localeName).map(normalizeLocale);
-    const current = get(this, '_locale');
+    let proposed = makeArray(localeName).map(normalizeLocale);
+    let current = get(this, '_locale');
 
     if (!isArrayEqual(proposed, current)) {
+      proposed.forEach(this.registerLocale, this);
       this.propertyWillChange('locale');
       set(this, '_locale', proposed);
       this.propertyDidChange('locale');
       this.trigger('localeChanged');
     }
   },
 
-  getFormat(formatType, format) {
-    const formats = get(this, 'formats');
+  /* @private */
+  lookupFormat(formatType, format) {
+    let formats = get(this, 'formats');
 
     if (formats && formatType && typeof format === 'string') {
       return get(formats, `${formatType}.${format}`);
     }
-
-    return {};
   },
 
+  /* @private */
   localeFactory(localeName) {
     return get(this, 'adapter').localeFactory(normalizeLocale(localeName), true);
   },
 
-  findTranslationByKey(key, localeName, options) {
-    return this.lookup(key, localeName, options);
+  /* @private */
+  registerLocale(localeName) {
```

Comment:
This is invoked when a localeName hasn't been seen by the service.  It will attempt to find and register the bundled translations and CLDR "on-demand" versus when the service is initialized (which involved abusing private APIs).